[cooper] 미션2 : 체스판 생성

pbg0205 opened the pull request:
안녕하세요.

2021 코드스쿼드 마스터즈 코스 백엔드의 Cooper 입니다.

미션2 기능 구현 pull request 부탁드립니다.

- Board 클래스에서 Pawn을 저장하기 위해 List를 선언하였습니다. 그리고 기본 생성자에서 ArrayList로 인스턴스화 하도록 구현하였습니다.
- Board 클래스를 테스트하기 위해 BoardTest 클래스를 정의했습니다. create 메서드를 통해 Board의 ArrayList 내에  저장 및 사이즈 확인 여부를 확인하는 테스트 코드를 작성했습니다.
- BoardTest 클래스에서 findPawn 메서드에서 발생할 수 있는 예외를 확인하는 테스트 코드를 작성했습니다.

pbg0205 committed:
feat : Pawn 클래스 기본 생성자 추가

- Pawn의 기본 생성자를 선언했습니다. Pawn의 기본 색은 흰색으로 설정되도록 구현했습니다.

pbg0205 committed:
feat : Board class 생성

- 폰을 찾기 위한 입력하는 인덱스 값과 List를 탐색하는 인덱스의 기준이
차이가 있기 때문에 findPawn 메서드 상단부에 index 변수를 선언했습니다.

- 추가한 폰의 깻수를 확인하기 위해 List 내 기본 메서드인 size 메서드를
그대로 인용하여 구현했습니다.

- index로 Pawn 탐색 시, index 범위 외 입력 값을 대비해 예외를  발생
하도록 구현했습니다. 추가적인 구현 시, 예외 처리를 진행할 예정입니다.

pbg0205 committed:
test : BoardTest 작성

- board에서 구현한 메서드(add, size, findPawn)을 확인하기 위해
흰색 폰, 검은 색 폰 인스턴스를 생성했습니다. 그리고 예상한 결과 값과
일치 여부를 확인하는 메서드인 assertEquals을  사용해 예상한 값과 일치함을
 확인했습니다.

- findPawn메서드에서 pawnList의 사이즈가 0일 때 예외를 확인하기 위해
@Test에 IllegalArgumentExeption을 추가했고 예상한 예외가 발생여부를
확인 했습니다.

- findPawn메서드에서 pawnList의 인덱스를 벗어난 입력에 대한 예외를
확인하기 위해 @Test에 IndexOutOfBoundsException을 추가했고 예상한
예외가 발생여부를 확인 했습니다.

pbg0205 committed:
fix : Board 클래스 findPawn 메서드 버그 수정

- index의 범위를 재설정하여 해당 예외가 적절하게 발생하도록
수정하였습니다.

pbg0205 committed:
test : BoardTest 테스트 코드 리팩토링

- mission01에 문자열 입력을 받는  Pawn 생성자로 변경하였습니다.
- assertThrows 메서드를 사용해서 예외 발생을 확인하는 테스트 코드를
작성했습니다.

SANGCO commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -0,0 +1,34 @@
+package cooper.chess;
```
net/honux/chess 패키지는 이제 지워도 될거 같네요~

SANGCO commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -0,0 +1,34 @@
+
+        return pawnList.get(index);
+    }
+}
```
![image](https://user-images.githubusercontent.com/26179444/107597154-2c8b0a00-6c5d-11eb-94c0-bc0300bf84ed.png)
IntelliJ 설정을 찾아서 적용해 보아요.
[How to make IntelliJ IDEA insert a new line at every end of file?](https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file)
[파일 끝에 개행을 추가해야 하는 이유](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/)

SANGCO commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -0,0 +1,49 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class BoardTest {
```
public 접근 제한자를 제거해도 좋을거 같아요.
JUnit5에서는 테스트 클래스나 메서드에 public을 붙이지 않아도 된답니다.

SANGCO commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -0,0 +1,49 @@
+        Pawn black = new Pawn("black");
+        board.add(black);
+        assertEquals(2, board.size());
+        assertEquals(black, board.findPawn(2));
```
assertAll()에 대해서 학습하고 적용해 보는건 어떨까요~
[JUnit 5 소개](https://javacan.tistory.com/entry/JUnit-5-Intro) (링크에 assertAll() 부분 참고해 주세요.)

SANGCO replied:
Color enum 만드셨던데 "black" 이렇게 문자열 쓰기보다 enum을 사용해 보아요.

SANGCO commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public Pawn findPawn(int number) {
+        int index = number - 1;
```
이 부분이 살짝 좀 그렇네요.
인덱스를 그대로 쓰는게 더 좋을거 같은데
일단은 이대로 한번 가보시죠~

honux77 replied:
매개변수 number가 무엇을 의미하는지 구체적이지 않아요.  매개변수명을 index로 사용하고 지역변수를 없애는게 저도 좋았을 것 같습니다.

SANGCO commented on src/test/java/cooper/chess/BoardTest.java:
```diff
@@ -0,0 +1,49 @@
+
+        assertThrows(IndexOutOfBoundsException.class,
+                () -> board.findPawn(-1));
+        assertThrows(IndexOutOfBoundsException.class,
+                () -> board.findPawn(2));
```
Board에 Pawn 2개 넣고 findPawn(2)하면 IndexOutOfBoundsException이 발생하나요?
```java
int index = number - 1;
if ((0 > index) || (index >= pawnList.size())) {
```
2가 들어가면 index는 1이 되고 `if ((0 > 1) || (1 >= 2))` 이렇게 되는게 아닌가 해서요~
제가 착각하는게 있으면 댓글로 좀 알려주세요~

honux77 replied:
여기에 좋아요와 해결됨만 있는데 뭐가 좋아요고 뭐가 해결됨인지 저도 궁금하네요. 쿠퍼 댓글로 좀 달아주세요.

SANGCO reviewed (commented):
cooper님 안녕하세요, 리뷰어 상코입니다.
작성하신 코드에 대해서 코멘트를 남겼습니다. 
확인해 보시고 질문 있으시면 댓글 부탁드립니다~

pbg0205 committed:
style : BoardTest 클래스 접근제한자 제거

- BoardTest 클래스의 public 접근제한자를 제거 했습니다.

pbg0205 committed:
test : BoardTest 클래스 테스트 코드 수정

- create 메서드의 테스트 코드를 assertAll 메서드를 사용해 각각의
테스트 코드를 확인할 수 있도록 수정했습니다.

- testOutOfBoardSize 메서드의 테스트 코드를 assertAll 메서드를 사용해
각각의 테스트 코드를 확인할 수 있도록 수정했습니다.

pbg0205 committed:
fix : Board 클래스 예외 발생 범위 변경

- findPawn 메서드에서 index 범위 외 발생하는 예외 범위를 수정 했습니다.

honux77 commented on src/main/java/cooper/chess/Board.java:
```diff
@@ -22,11 +22,11 @@ public Pawn findPawn(int number) {
 
         if(pawnList.size() == 0) {
-            throw new IndexOutOfBoundsException("board size == 0입니다.");
+            throw new IndexOutOfBoundsException("사이즈가 0 입니다.");
```
사이즈가 0일때 IndexOutOfBoundsException을 내는 건 맞지 않은 것 같아요. 
아래도 마찬가지입니다. 
예외를 던지고 싶다면 어떤 종류의 예외가 적당한지 한번 생각해 보세요.

honux77 commented:
수고하셨습니다.  큐퍼는 리뷰어와 조금 더 적극적으로 커뮤니케이션을 하면 좋을 것 같아요! 화이팅!

pbg0205 commented:
@honux77 리뷰가 아직 적응이 되지 않아 기본적인 실수를 많이 하게 되었습니다 . 빠르게 적응할 수 있도록 신경쓰겠습니다.
앞으로도 부족한 부분 있으면 거침없는 피드백 부탁드립니다! 감사합니다!
